pyccel/ast/builtins.py

```diff
@@ -245,49 +181,73 @@ def fprint(self, printer):
         return code
 
 #==============================================================================
-class PythonTuple(Function):
+class PythonTuple(Expr, PyccelAstNode):
     """ Represents a call to Python's native tuple() function.
     """
-    _iterable = True
-    _arg_dtypes = None
-    _is_homogeneous = False
-    is_zero = False
-    def __new__(cls, args):
-        if not iterable(args):
-            args = [args]
-        args = tuple(args)
+    _iterable        = True
+    _is_homogeneous  = False
 
+    def __new__(cls, *args):
         return Basic.__new__(cls, *args)
 
-    @property
-    def dtype(self):
-        return 'tuple'
-
-    @property
-    def homogeneous_dtype(self):
-        assert(self.is_homogeneous)
-        return self._homogeneous_dtype
-
-    @property
-    def shape(self):
-        if(self._arg_dtypes is None):
-            return [Integer(len(self._args))]
+    def __init__(self, *args, **kwargs):
+        if self.stage == 'syntactic' or len(args) == 0:
+            return
+        is_homogeneous = all(args[0].dtype==a.dtype for a in args[1:])
+        is_homogeneous = is_homogeneous and all(args[0].rank==a.rank   for a in args[1:])
```

Comment:
yes all the arguments must have the same type, shape and rank
shape cannot be always checked so I leave that for the fortran compiler to raise an error

Reply:
Ah, sorry I meant to write shape

Reply:
The shape must always be a known value or a PyccelArraySize right?
I think it would be better to make the tuple unhomogeneous if the sizes aren't the same or aren't known

Reply:
if the user writes a function like
```
def f(x):
    y = (x[0], x[1])
```
it will be inhomogeneous because we don't know the shape of `x`, I think it's better to let it be homogeneous by default and the exception should be inhomogeneous

Reply:
Maybe we can discuss this this afternoon.
Personally I disagree. If the user wanted something contiguous then they would have probably written:
```
def f(x):
    y = x[0:2]
```
Tuples are usually used for inhomogeneous data. We have a specialised case for homogeneous data to improve performance but if we make homogeneous the default then the behaviour will be broken in cases with different shapes without there being an alternative way to write the same thing. Whereas there is an alternative way to write the homogeneous case